### PR TITLE
Improved SQL Query performance by optimizing getter caching.

### PR DESCRIPTION
**Pull request by pveentjer**
Instead of doing a string concatenation of class + attributename, use a map of maps with class as first key and attribute name as second.

Performance improvement:

before:
596.072222222

after
663.924444444

This is measured using the following stabilizer test on our testlab:
```
class=com.hazelcast.stabilizer.tests.map.SqlPredicateTest
threadCount = 1
keyLength = 10
keyCount = 10000
logFrequency = 10000
performanceUpdateFrequency = 1
basename = ObjectMap
sql = age = -1 
probe-latency = latency
intervalMs = 0
```

Using the following HZ configuration
```
    <map name="ObjectMap*">
        <in-memory-format>OBJECT</in-memory-format>
    </map>
```

Of course is depends a lot on the actual query, number of entries etc etc etc.

When I directly make use of mapContextQuerySupport.queryOnPartition the performance improvement is a lot more impressive:
before:
6734.32421875 queries/second
after:
19037.46484375 queries/second

### CODE REVIEW

File: hazelcast/src/main/java/com/hazelcast/query/impl/ReflectionHelper.java [hunk 1]
```diff
@@ -62,8 +66,6 @@ public static AttributeType getAttributeType(Class klass) {
             return AttributeType.BIG_INTEGER;
         } else if (klass == float.class || klass == Float.class) {
             return AttributeType.FLOAT;
-        } else if (klass == short.class || klass == Short.class) {
-            return AttributeType.SHORT;
```
Comment (ahmetmircik) [inline hazelcast/src/main/java/com/hazelcast/query/impl/ReflectionHelper.java]: what is the point of removing `SHORT` part? Can you put a note?

File: hazelcast/src/main/java/com/hazelcast/query/impl/ReflectionHelper.java [hunk 2]
```diff
@@ -82,6 +84,28 @@ public static AttributeType getAttributeType(Class klass) {
...
+        return cache.get(attribute);
+    }
+
+    private static Getter set(Class clazz, String attribute, Getter getter) {
+        ConcurrentMap<String, Getter> cache = GETTER_CACHE.get(clazz);
+        if (cache == null) {
+            cache = new ConcurrentHashMap<String, Getter>();
+            ConcurrentMap<String, Getter> replaced = GETTER_CACHE.putIfAbsent(clazz, cache);
+            cache = replaced == null ? cache : replaced;
+        }
```
Comment (ahmetmircik) [inline hazelcast/src/main/java/com/hazelcast/query/impl/ReflectionHelper.java]: Seems replaceable with `ConcurrencyUtil#`getOrPutIfAbsent` ?

File: hazelcast/src/main/java/com/hazelcast/query/impl/ReflectionHelper.java [hunk 3]
```diff
@@ -91,13 +115,15 @@ public static AttributeType getAttributeType(Object value, String attribute) {
...
-        final String cacheKey = clazz.getName() + ":" + attribute;
-        Getter getter = GETTER_CACHE.get(cacheKey);
+        final Class targetClazz = obj.getClass();
+        Class clazz = targetClazz;
+        Getter getter = get(clazz, attribute);
         if (getter != null) {
             return getter;
         }
 
+        System.out.println("extracting getter: " + obj.getClass().getSimpleName() + "." + attribute);
```
Comment (ahmetmircik) [inline hazelcast/src/main/java/com/hazelcast/query/impl/ReflectionHelper.java]: Is `System.out` intentional?

File: hazelcast/src/main/java/com/hazelcast/query/impl/ReflectionHelper.java [hunk 4]
```diff
@@ -193,10 +217,14 @@ public Getter(final Getter parent) {
 
         Object getValue(Object obj) throws Exception {
             Object paramObj = obj;
+            if (parent != null) {
+                System.out.println("unexpected parent");
```
Comment (ahmetmircik) [inline hazelcast/src/main/java/com/hazelcast/query/impl/ReflectionHelper.java]: Is `System.out` intentional?

File: hazelcast/src/main/java/com/hazelcast/query/impl/ReflectionHelper.java (hunk 1 above)
Comment (pveentjer) [inline hazelcast/src/main/java/com/hazelcast/query/impl/ReflectionHelper.java]: Hmmm.. why did that get removed????

File: hazelcast/src/main/java/com/hazelcast/query/impl/ReflectionHelper.java (hunk 3 above)
Comment (pveentjer) [inline hazelcast/src/main/java/com/hazelcast/query/impl/ReflectionHelper.java]: Must have been sleeping.

File: hazelcast/src/main/java/com/hazelcast/query/impl/ReflectionHelper.java (hunk 4 above)
Comment (pveentjer) [inline hazelcast/src/main/java/com/hazelcast/query/impl/ReflectionHelper.java]: This one was already there. I think we should get rid of it.

**Comment by pveentjer** [comment]
Thanks!

**Pull request merged**